File: shcloj2/code/src/examples/multimethods/account.clj
;;
;;Excerpted from "Programming Clojure, Second Edition",
;;published by The Pragmatic Bookshelf.
;;Copyrights apply to this code. It may not be used to create training material, 
;;courses, books, articles, and the like. Contact us if you are in doubt.
;;We make no guarantees that this code is fit for any purpose. 
;;Visit http://www.pragmaticprogrammer.com/titles/shcloj2 for more book information.
;;
(ns examples.multimethods.account)

(defstruct account :id :tag :balance)

(alias 'acc 'examples.multimethods.account)

(defmulti interest-rate :tag)
(defmethod interest-rate ::acc/Checking [_] 0M)
(defmethod interest-rate ::acc/Savings [_] 0.05M)

(defmulti account-level :tag)
(defmethod account-level ::acc/Checking [acct]
  (if (>= (:balance acct) 5000) ::acc/Premium ::acc/Basic))
(defmethod account-level ::acc/Savings [acct]
  (if (>= (:balance acct) 1000) ::acc/Premium ::acc/Basic))

File: hwcuc/code/databases/00/lib/account.rb
#---
# Excerpted from "The Cucumber Book",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material, 
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose. 
# Visit http://www.pragmaticprogrammer.com/titles/hwcuc for more book information.
#---
require 'active_record'
ActiveRecord::Base.establish_connection(:adapter => 'sqlite3', 
                                        :database => 'db/bank.db')
ActiveRecord::Migrator.migrate("db/migrate/")
class Account < ActiveRecord::Base
  validates_uniqueness_of :number
  def queue
    @queue ||= TransactionQueue.new
  end
  def credit(amount)
    queue.write("+#{amount},#{number}")
  end
  def debit(amount)
    queue.write("-#{amount},#{number}")
  end
end

